prow/cmd/build/controller.go

```diff
@@ -17,46 +17,424 @@ limitations under the License.
+	stop := make(chan struct{})
+	c := make(chan os.Signal, 2)
+	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
+	go func() {
+		<-c
+		close(stop)
+		<-c
+		os.Exit(1)
+	}()
+	return stop
+}
+
 func main() {
-	logrus.Fatalf("see next pr")
+	flag.Parse()
+	logrusutil.NewDefaultFieldsFormatter(nil, logrus.Fields{"component": "build"})
+	stop := stopper()
+
+	cfg, err := clientcmd.BuildConfigFromFlags(masterURL, kubeconfig)
+	if err != nil {
+		logrus.Fatalf("Error building kubeconfig: %v", err)
+	}
+
+	kc, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		logrus.Fatalf("Error building kubernetes client: %v", err)
+	}
+
+	pjc, err := prowjobset.NewForConfig(cfg)
+	if err != nil {
+		logrus.Fatalf("Error building prowjob client: %v", err)
+	}
+
+	bc, err := buildset.NewForConfig(cfg)
+	if err != nil {
+		logrus.Fatalf("Error building build client: %v", err)
+	}
+
+	// Assume watches receive updates, but resync every 30m in case something wonky happens
+	pjif := prowjobinfo.NewSharedInformerFactory(pjc, time.Minute*30)
+	bif := buildinfo.NewSharedInformerFactory(bc, time.Minute*30)
+
+	controller := newController(kc, pjc, bc, pjif.Prow().V1().ProwJobs(), bif.Build().V1alpha1().Builds())
+
+	go pjif.Start(stop)
+	go bif.Start(stop)
+
+	if err = controller.Run(2, stop); err != nil {
+		logrus.Fatalf("Error running controller: %v", err)
+	}
+}
+
+type controller struct {
+	pjc prowjobset.Interface
+	bc  buildset.Interface
+
+	pjLister prowjoblisters.ProwJobLister
+	pjSynced cache.InformerSynced
+	bLister  buildlisters.BuildLister
+	bSynced  cache.InformerSynced
+
+	workqueue workqueue.RateLimitingInterface
+
+	recorder record.EventRecorder
+}
+
+func newController(kc kubernetes.Interface, pjc prowjobset.Interface, bc buildset.Interface, pji prowjobinfov1.ProwJobInformer, bi buildinfov1alpha1.BuildInformer) *controller {
+	// Log to events
+	prowjobscheme.AddToScheme(scheme.Scheme)
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logrus.Infof)
+	eventBroadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: kc.CoreV1().Events("")})
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, untypedcorev1.EventSource{Component: controllerName})
+
+	// Create struct
+	c := &controller{
+		pjc:       pjc,
+		bc:        bc,
+		pjLister:  pji.Lister(),
+		pjSynced:  pji.Informer().HasSynced,
+		bLister:   bi.Lister(),
+		bSynced:   bi.Informer().HasSynced,
+		workqueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName),
+		recorder:  recorder,
+	}
+
+	logrus.Info("Setting up event handlers")
+
+	// Reconcile whenever a prowjob changes
+	pji.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.enqueueKey,
+		UpdateFunc: func(old, new interface{}) {
+			c.enqueueKey(new)
+		},
+		DeleteFunc: c.enqueueKey,
+	})
+
+	// Reconvile whenever a build changes.
+	bi.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.enqueueKey,
```

Comment:
Is this assuming the build's name is also a ProwJob name? We should use an OwnerRef or something and do the resolution of which ProwJob to enqueue explicitly